nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/clustered/server/StandardLoadBalanceProtocol.java

```diff
@@ -0,0 +1,578 @@
+        }
+
+        logger.debug("Successfully received {} FlowFiles from Peer {} to Load Balance for Connection {}", flowFilesReceived.size(), peerDescription, connectionId);
+    }
+
+    private void completeTransaction(final InputStream in, final OutputStream out, final String peerDescription, final List<RemoteFlowFileRecord> flowFilesReceived,
+                                     final String nodeName, final String connectionId, final long startTimestamp, final LoadBalancedFlowFileQueue flowFileQueue) throws IOException {
+        final int completionIndicator = in.read();
+        if (completionIndicator < 0) {
+            throw new EOFException("Expected to receive a Transaction Completion Indicator from Peer " + peerDescription + " but encountered EOF");
+        }
+
+        if (completionIndicator == ABORT_TRANSACTION) {
+            throw new TransactionAbortedException("Peer " + peerDescription + " chose to Abort Load Balance Transaction");
+        }
+
+        if (completionIndicator != COMPLETE_TRANSACTION) {
+            logger.debug("Expected to receive Transaction Completion Indicator from Peer " + peerDescription + " but instead received a value of " + completionIndicator + ". Sending back an Abort " +
+                            "Transaction Flag.");
+            out.write(ABORT_TRANSACTION);
+            out.flush();
+            throw new IOException("Expected to receive Transaction Completion Indicator from Peer " + peerDescription + " but instead received a value of " + completionIndicator);
+        }
+
+        logger.debug("Received Complete Transaction indicator from Peer {}", peerDescription);
+        registerReceiveProvenanceEvents(flowFilesReceived, nodeName, connectionId, startTimestamp);
+        updateFlowFileRepository(flowFilesReceived, flowFileQueue);
+        transferFlowFilesToQueue(flowFilesReceived, flowFileQueue);
+
+        out.write(CONFIRM_COMPLETE_TRANSACTION);
+        out.flush();
+    }
+
+    private void registerReceiveProvenanceEvents(final List<RemoteFlowFileRecord> flowFiles, final String nodeName, final String connectionId, final long startTimestamp) {
+        final long duration = System.currentTimeMillis() - startTimestamp;
+
+        final List<ProvenanceEventRecord> events = new ArrayList<>(flowFiles.size());
+        for (final RemoteFlowFileRecord remoteFlowFile : flowFiles) {
+            final ProvenanceEventRecord provenanceEvent = new StandardProvenanceEventRecord.Builder()
+                    .fromFlowFile(remoteFlowFile.getFlowFile())
+                    .setEventType(ProvenanceEventType.RECEIVE)
+                    .setTransitUri("nifi://" + nodeName + "/loadbalance/" + connectionId)
+                    .setSourceSystemFlowFileIdentifier(remoteFlowFile.getRemoteUuid())
+                    .setEventDuration(duration)
+                    .setComponentId(connectionId)
+                    .setComponentType("Load Balanced Connection")
+                    .build();
```

Comment:
![image](https://user-images.githubusercontent.com/1107620/46131222-1770d480-c276-11e8-998c-c4a77853276d.png)

* Although the provenance event is created from a FlowFile, and the FlowFile's content size is set by content claim triple, somehow the provenance shows size as 0.
* `RECEIVE` is reported by `Load Balanced Connection`. I'd expect to the same for the `SEND` event, too. Is this intentional?

Reply:
It is not intentional. Will address.